### PR TITLE
Start a proper ledger API

### PR DESCRIPTION
**Pull request by lehins**
This is a very beginning of the `cardano-ledger-api`

For now it only includes one function: `setBabbageMinTxOut`

### CODE REVIEW

File: cabal.project [hunk 1]
```diff
@@ -17,6 +17,7 @@ packages:
   eras/shelley/test-suite
   eras/shelley-ma/impl
   eras/shelley-ma/test-suite
+  libs/cardano-ledger-api
```
Comment (teodanciu) [inline cabal.project:20]: Was wondering if  we want to add it to the list [below](https://github.com/input-output-hk/cardano-ledger/pull/2995/commits/1be5a8b04a35deeb75661f7d8862f3a5e2adb752#diff-b8ed06757045fac949c89f2139a862498ad2b6d1f82c61a31e7c91d6cf0eaa70L125) which enumerates each package in order to set " -Werror"  , but then I clicked on the link explaining why it's done and the github issue is closed, so maybe the whole enumerating is not necessary at all.  I can check later if that's the case

File: CHANGELOG.md [hunk 2]
```diff
@@ -158,8 +167,6 @@ BabbageEraTxBody --> AlonzoEraTxBody --> ....
       - `FromAlonzoUtxoFail` -> `AlonzoInBabbageUtxoPredFailure`
     - `BabbageUtxowPred` -> `BabbageUtxowPredFailure` and constructor:
       - `FromAlonzoUtxowFail` -> `AlonzoInBabbageUtxowPredFailure`
-- Changed `mint` field type to `MultiAsset (Crypto era)` in `MATxBody`, `AlonzoTxBody`, `BabbageTxBody`
```
Comment (teodanciu) [inline CHANGELOG.md:165]: Oops, sorry about that

File: libs/cardano-ledger-api/cardano-ledger-api.cabal [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+description:
+  This package a public interface into the cardano ledger codebase. It also
+  provides functionality that works for all eras and is useful for downstream
+  packages.
+bug-reports:        https://github.com/input-output-hk/cardano-ledger/issues
+
+license:            Apache-2.0
+author:             IOHK Formal Methods Team
+maintainer:         formal.methods@iohk.io
+copyright:          2021 Input Output (Hong Kong) Ltd.
```
Comment (teodanciu) [inline libs/cardano-ledger-api/cardano-ledger-api.cabal:14]: Do we ever update these? I see some are 2020 in the code :laughing:  But probably it doesn't matter

File: libs/cardano-ledger-api/cardano-ledger-api.cabal [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  exposed-modules:
+    Cardano.Ledger.Api.Tx
+    Cardano.Ledger.Api.Tx.Out
+
+  build-depends:
+    cardano-ledger-core,
+    cardano-ledger-alonzo,
+    cardano-ledger-babbage,
+    cardano-ledger-shelley,
+    microlens,
```
Comment (teodanciu) [inline libs/cardano-ledger-api/cardano-ledger-api.cabal]: Also not  important nitpick: a new line might be nice here

File: cabal.project (hunk 1 above)
Comment (lehins) [reply cabal.project:20]: Yeah, you are right on both counts. We do need to add a line for that package (took care of that in this PR). And yes it was fixed, but we need to upgrade to cabal-3.8 for a single top level `-Werror` to work. I've created a ticket for that: https://github.com/input-output-hk/cardano-ledger/issues/2998

File: libs/cardano-ledger-api/cardano-ledger-api.cabal (hunk 3 above)
Comment (lehins) [reply libs/cardano-ledger-api/cardano-ledger-api.cabal:14]: I think it should matter. I created a ticket about this as well: https://github.com/input-output-hk/cardano-ledger/issues/2999

File: CHANGELOG.md (hunk 2 above)
Comment (lehins) [reply CHANGELOG.md:165]: No worries :) You did it right. It was after your PR was merged that the header was just switched from `### Unreleased` to `### 1.1.x`. Not your fault at all.

File: libs/cardano-ledger-api/cardano-ledger-api.cabal (hunk 4 above)
Comment (lehins) [reply libs/cardano-ledger-api/cardano-ledger-api.cabal]: Good nitpick. Done

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/Tx/Out.hs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  Sized (TxOut era) ->
+  Sized (TxOut era)
+setBabbageMinTxOut pp = go
+  where
+    go txOut =
+      let curMinCoin = babbageMinUTxOValue pp txOut
+          curCoin = txOut ^. toSizedL coinTxOutL
+       in if curCoin == curMinCoin
+            then txOut
+            else go (txOut & toSizedL coinTxOutL .~ curMinCoin)
```
Comment (teodanciu) [inline libs/cardano-ledger-api/src/Cardano/Ledger/Api/Tx/Out.hs:47]: I've been staring at this line for a long time - am I right: 
It's invoking the function recursively with the parameter set to txOut modified to have the coin set to the new `curMinCoin`, correct?

File: libs/cardano-ledger-api/cardano-ledger-api.cabal [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+version:            0.1.0.0
+synopsis:           Public API for the cardano ledger codebase
+description:
+  This package a public interface into the cardano ledger codebase. It also
+  provides functionality that works for all eras and is useful for downstream
+  packages.
+bug-reports:        https://github.com/input-output-hk/cardano-ledger/issues
+
+license:            Apache-2.0
+author:             IOHK Formal Methods Team
```
Comment (JaredCorduan) [inline libs/cardano-ledger-api/cardano-ledger-api.cabal:12]:
```suggestion
author:             IOG Ledger Team
```

Using "Formal Methods" is probably left over from when this was the ledger specs repo. I _think_ IOG Ledger Team is better.

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/Tx.hs [hunk 7]
```diff
@@ -0,0 +1,16 @@
+module Cardano.Ledger.Api.Tx
+  ( EraTx (..),
+
+    -- * Shelley Era
+    ShelleyTx,
+
```
Comment (JaredCorduan) [inline libs/cardano-ledger-api/src/Cardano/Ledger/Api/Tx.hs:6]: What about Allegra and Mary?

**Review by teodanciu — APPROVED** [review]
Looks good to me.

Exciting to have an API we can look after properly!

**Comment by teodanciu** [comment]
Was just gonna say  @JaredCorduan will have more insight to review, and I was not wrong it seems!

File: libs/cardano-ledger-api/cardano-ledger-api.cabal (hunk 6 above)
Comment (lehins) [reply libs/cardano-ledger-api/cardano-ledger-api.cabal:12]: This was copy and paster. I've created a ticket for that: https://github.com/input-output-hk/cardano-ledger/issues/2999

All cabal files suffer from this, so it is best to take care of this as one task. So, if you can comment on that ticket what the proper values should be that will be great.

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/Tx/Out.hs (hunk 5 above)
Comment (lehins) [reply libs/cardano-ledger-api/src/Cardano/Ledger/Api/Tx/Out.hs:47]: Correct.

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/Tx.hs (hunk 7 above)
Comment (lehins) [reply libs/cardano-ledger-api/src/Cardano/Ledger/Api/Tx.hs:6]: They've got no `Tx` type

File: libs/cardano-ledger-api/cardano-ledger-api.cabal (hunk 6 above)
Comment (JaredCorduan) [reply libs/cardano-ledger-api/cardano-ledger-api.cabal:12]: perfect, thanks!

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/Tx.hs (hunk 7 above)
Comment (JaredCorduan) [reply libs/cardano-ledger-api/src/Cardano/Ledger/Api/Tx.hs:6]: ah yes, just the body!

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/Tx/Out.hs (hunk 5 above)
Comment (JaredCorduan) [reply libs/cardano-ledger-api/src/Cardano/Ledger/Api/Tx/Out.hs:47]: could we add a test of some kind, just to demonstrate that this function does indeed converge?

**Review by JaredCorduan — APPROVED** [review]
long overdue and very exciting!! thanks for starting this @lehins , this is a great start

File: libs/cardano-ledger-api/src/Cardano/Ledger/Api/Tx/Out.hs (hunk 5 above)
Comment (lehins) [reply libs/cardano-ledger-api/src/Cardano/Ledger/Api/Tx/Out.hs:47]: > could we add a test of some kind, just to demonstrate that this function does indeed converge?

Very good idea. I created a new PR for that: https://github.com/input-output-hk/cardano-ledger/pull/3004

**Pull request merged**